File: Cpp/931_Minimum_Falling_Path_Sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int result = INT_MAX;
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        
        for(int i = 0; i < n; ++i) {
            dp[0][i] = matrix[0][i];
        }
        
        for(int i = 1; i < m; ++i) {
            for(int j = 0; j < n; ++j) {
                int path = dp[i-1][j];
                if(j > 0) {
                    path = min(path, dp[i-1][j-1]);
                }
                
                if(j < n-1) {
                    path = min(path, dp[i-1][j+1]);
                }
                
                dp[i][j] = matrix[i][j] + path;
            }
        }
        
        for(int i = 0; i < n; ++i) {
            result = min(result, dp[m-1][i]);
        }
        
        return result;
    }
};

File: Kotlin/931_Minimum_Falling_Path_Sum.kt
class Solution {
    fun minFallingPathSum(matrix: Array<IntArray>): Int {
        var result = Integer.MAX_VALUE
        val m = matrix.size
        val n = matrix[0].size
        val dp = Array(m){ IntArray(n) {0} }
        
        for(i in 0 until n) {
            dp[0][i] = matrix[0][i]
        }
        
        for(i in 1 until m) {
            for(j in 0 until n) {
                var path = dp[i - 1][j]
                if(j > 0) {
                    path = minOf(path, dp[i - 1][j - 1])
                }
                
                if(j < n - 1) {
                    path = minOf(path, dp[i - 1][j + 1])
                }
                
                dp[i][j] = matrix[i][j] + path
            }
        }
        
        for(i in 0 until n) {
            result = minOf(result, dp[m - 1][i])
        }
        
        return result
    }
}